src/contacts/contacts.py

```diff
@@ -26,208 +26,172 @@
+    def add_contact(username):
+        """Add a new favorite account to user's contacts list
+
+        Fails if account or routing number are invalid
+        or if label is not alphanumeric
+
+        request fields:
+        - account_num
+        - routing_num
+        - label
+        - is_external
+        """
+        auth_header = request.headers.get("Authorization")
+        if auth_header:
+            token = auth_header.split(" ")[-1]
+        else:
+            token = ""
+        try:
+            auth_payload = jwt.decode(
+                token, key=app.config["PUBLIC_KEY"], algorithms="RS256"
+            )
+            if username != auth_payload["user"]:
+                raise PermissionError
+            req = {
+                k: (bleach.clean(v) if isinstance(v, str) else v)
+                for k, v in request.get_json().items()
+            }
+            _validate_new_contact(req)
+
+            _check_contact_allowed(username, auth_payload["acct"], req)
+            # Create contact data to be added to the database.
+            contact_data = {
+                "username": username,
+                "label": req["label"],
+                "account_num": req["account_num"],
+                "routing_num": req["routing_num"],
+                "is_external": req["is_external"],
+            }
+            # Add contact_data to database
+            contacts_db.add_contact(contact_data)
+            return jsonify({}), 201
+
+        except (PermissionError, jwt.exceptions.InvalidTokenError):
+            return jsonify({"msg": "authentication denied"}), 401
+        except UserWarning as warn:
+            return jsonify({"msg": str(warn)}), 400
+        except ValueError as err:
+            return jsonify({"msg": str(err)}), 409
+        except SQLAlchemyError as err:
+            app.logger.error(err)
+            return jsonify({"msg": "failed to add contact"}), 500
+
+    def _validate_new_contact(req):
+        """Check that this new contact request has valid fields"""
+        app.logger.debug("validating add contact request: %s", str(req))
+        # Check if required fields are filled
+        fields = ("label", "account_num", "routing_num", "is_external")
+        if any(f not in req for f in fields):
+            raise UserWarning("missing required field(s)")
+
+        # Validate account number (must be 10 digits)
+        if not re.match(r"\A[0-9]{10}\Z", req["account_num"]):
+            raise UserWarning("invalid account number")
+        # Validate routing number (must be 9 digits)
+        if not re.match(r"\A[0-9]{9}\Z", req["routing_num"]):
+            raise UserWarning("invalid routing number")
+        # Only allow external accounts to deposit
+        if (req["is_external"] and req["routing_num"] == app.config["LOCAL_ROUTING"]):
+            raise UserWarning("invalid routing number")
+        # Validate label
+        # Must be >0 and <30 chars, alphanumeric and spaces, can't start with space
```

Comment:
Actually I think this might be <= 30? Since the first character counts too

Reply:
done